### PR TITLE
[Windows Upgrade] EPIC D2IQ-61891 : Support Windows Node Update - improved

### PR DESCRIPTION
**Pull request by sergiimatusEPAM**
There is set of JIRAs under epic D2IQ-61891 : Support Windows Node Update which covers:
- templatizing dcos_install.ps1 and adding bootstrap_url/masters as gen_out arguments.
- pass generation of dcos_install.ps1 to gen.build_deploy.powershell.py
- created a template for dcos_node_upgrade and similarly to upgrade.py such put into upgrade_win.py
- generation of dcos_node_upgrade.py using upgrade_win.py

## High-level description

What features does this change enable? What bugs does this change fix?

## Corresponding DC/OS tickets (required)

  - [D2IQ-61891](https://jira.d2iq.com/browse/D2IQ-61891) EPIC to Support Windows Node Update (Jon Giddy)

## Related tickets (optional)
  - [D2IQ-64278](https://jira.d2iq.com/browse/D2IQ-64278) Fill up functionality requires in dcos_node_upgrade.ps1/.py (Vadym Zagorodko)
  - [D2IQ-64275](https://jira.d2iq.com/browse/D2IQ-64275) Create gen/build_deploy/powershell.py (Sergii Matus)
  - [D2IQ-64277](https://jira.d2iq.com/browse/D2IQ-64277) Move bootstrap.url along with a Masters list into powershell.py (Sergii Matus)
  - [D2IQ-64280](https://jira.d2iq.com/browse/D2IQ-64280) Integrate upgrade_win.py (Sergii Matus)
Bugfix : moved dcos_node_upgrade to template instead of .py in-line str (Sergii Matus)
Bugfix : workaround to check a 'master_discovery' value prior filling up 'master_list' value (Sergii Matus)

  - [DCOS-ID](https://jira.mesosphere.com/browse/DCOS-<number>) JIRA title / short description.

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain.
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]

### CODE REVIEW

**Comment by sergiimatusEPAM** [comment]
@mesosphere-mergebot changelog-not-required JIRAs of Windows Upgrade scope

**Comment by sergiimatusEPAM** [comment]
@mesosphere-mergebot label Ready For Review

**Comment by sergiimatusEPAM** [comment]
@mesosphere-mergebot request-review @jongiddy

**Comment by sergiimatusEPAM** [comment]
@mesosphere-mergebot request-review @DominikDary

File: dcos_installer/cli.py [hunk 1]
```diff
@@ -141,6 +145,15 @@ def get_argument_parser():
         help='Generate a script that upgrades DC/OS nodes running installed_cluster_version'
     )
 
+    mutual_exc.add_argument(
```
Comment (jongiddy) [inline dcos_installer/cli.py]: Do we need a separate parameter here? Why do we not generate the Windows upgrade script at the same time that we generate the Linux upgrade script?

File: gen/build_deploy/powershell.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    if gen_out.arguments['master_discovery'] == 'static':
+        master_list = gen_out.arguments['master_list']
+    elif gen_out.arguments['master_discovery'] == 'master_http_loadbalancer':
+        master_list = gen_out.arguments['exhibitor_address'] + ':2181'
+    else:
+        master_list = 'zk-1.zk:2181,zk-2.zk:2181,zk-3.zk:2181,zk-4.zk:2181,zk-5.zk:2181'
+
+    powershell_template_path = os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                                            'powershell/dcos_install.ps1.in')
+    with open(powershell_template_path, 'r') as f:
```
Comment (jongiddy) [inline gen/build_deploy/powershell.py]: :+1: I much prefer reading in a template to embedding the template in the Python code. Not in this PR, but we should change the Linux gen to do this too.

File: gen/build_deploy/powershell/dcos_node_upgrade.ps1.in [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        [ScriptBlock] $ScriptBlock,
+        [int[]] $AllowedExitCodes = @(0),
+        [switch] $IgnoreExitCode
+    )
+    [string] $stringScriptBlock = $ScriptBlock.ToString();
+    $backupErrorActionPreference = $ErrorActionPreference;
+    $ErrorActionPreference = "Continue";
+    try
+    {
+        if (Test-CalledFromPrompt)
```
Comment (jongiddy) [inline gen/build_deploy/powershell/dcos_node_upgrade.ps1.in]: What am I missing? The two branches here appear to be the same.

File: dcos_installer/cli.py (hunk 1 above)
Comment (sergiimatusEPAM) [reply dcos_installer/cli.py]: well, currently the generation of dcos_node_upgrade.ps1 has been separated by the cli's arg as diff function must be used and slightly diff routines happens to generate the ps1 script. 
If you think, that we should combine such under --generate-node-upgrade-script, it's also doable

File: gen/build_deploy/powershell/dcos_node_upgrade.ps1.in (hunk 3 above)
Comment (sergiimatusEPAM) [reply gen/build_deploy/powershell/dcos_node_upgrade.ps1.in]: Fixed by removing stderr to stdout output if a command run in a subcmd (prompt), as it's redundant append. Tested

File: dcos_installer/cli.py (hunk 1 above)
Comment (jongiddy) [reply dcos_installer/cli.py]: OK, let's leave it for now, and then look at how this gets used.

**Pull request merged**